chain/chunks/src/lib.rs

```diff
@@ -1288,48 +1396,41 @@ impl ShardsManager {
             }
         }
 
-        // 4. Checking epoch_id validity
-        let prev_block_hash = header.prev_block_hash();
-        let epoch_id = match self.runtime_adapter.get_epoch_id_from_prev_block(&prev_block_hash) {
-            Ok(epoch_id) => epoch_id,
-            Err(_) => {
-                // It may happen because PartialChunkEncodedMessage appeared before Block announcement.
-                // We keep the chunk until Block is received.
-                return Ok(ProcessPartialEncodedChunkResult::NeedBlock);
-            }
-        };
+        // 1.c checking header validity
+        match partial_encoded_chunk
+            .validate_with(|pec| self.validate_chunk_header(chain_head, &pec.header).map(|()| true))
+        {
+            Err(Error::ChainError(chain_error)) => match chain_error.kind() {
+                ErrorKind::DBNotFoundErr(_) => {
+                    return Ok(ProcessPartialEncodedChunkResult::NeedBlock)
+                }
+                _ => return Err(chain_error.into()),
+            },
+            Err(err) => return Err(err),
+            Ok(_) => (),
+        }
+        let partial_encoded_chunk = partial_encoded_chunk.into_inner();
 
-        // 5. Checking part_ords' validity
+        // 1.d Checking part_ords' validity
         let num_total_parts = self.runtime_adapter.num_total_parts();
         for part_info in partial_encoded_chunk.parts.iter() {
             // TODO: only validate parts we care about
             self.validate_part(header.encoded_merkle_root(), part_info, num_total_parts)?;
         }
 
-        // 6. Checking receipts validity
+        // 1.e Checking receipts validity
         for proof in partial_encoded_chunk.receipts.iter() {
+            // TODO: only validate receipts we care about
             let shard_id = proof.1.to_shard_id;
-            if self.cares_about_shard_this_or_next_epoch(
-                self.me.as_ref(),
-                &prev_block_hash,
-                shard_id,
-                true,
-            ) {
```

Comment:
because `prev_block_hash` is not ready and technically we don't know if we track this shard or not. Considering now every node tracks all shard, I think it's not a big issue. I'll create a TODO to track this

Reply:
https://github.com/near/nearcore/issues/5885